File: python/example/simple_mapper.py
from pathlib import Path
from typing import Any, override
import numpy as np
import json
import libtokamap


def dot_product(inputs: dict[str, np.array], _params: dict[str, Any]) -> np.array:
    if 'lhs' not in inputs:
        raise ValueError("lhs is required")
    if 'rhs' not in inputs:
        raise ValueError("rhs is required")
    lhs = inputs['lhs']
    rhs = inputs['rhs']
    if lhs.size != rhs.size:
        raise ValueError("lhs and rhs must have the same size")
    if lhs.ndim != 1 or rhs.ndim != 1:
        raise ValueError("lhs and rhs must be 1-dimensional")
    return np.array(np.dot(lhs, rhs))


class JSONDataSource(libtokamap.DataSource):
    def __init__(self, data_root: Path):
        self.data_root = data_root

    @override
    def get(self, args: dict[str, str]) -> np.ndarray:
        if 'file_name' not in args:
            raise ValueError("file_name is required")
        if 'signal' not in args:
            raise ValueError("signal is required")

        file_name = args['file_name']
        signal = args['signal']

        with open(self.data_root / file_name, 'r') as f:
            data = json.load(f)

        tokens = signal.split('/')
        for token in tokens:
            try:
                num = int(token)
                data = data[num]
            except ValueError:
                data = data[token]

        return np.array(data)


def map(mapper: libtokamap.Mapper, mapping: str, signal: str):
    res = mapper.map(mapping, signal, {'shot': 42})
    if res.dtype == 'S1':
        res = res.tobytes().decode()
    print(f"{signal}: {res}")
    return res


def map_all(mapper: libtokamap.Mapper, mapping: str):
    map(mapper, mapping, "magnetics/version")
    n_coils = map(mapper, mapping, "magnetics/coil")
    for coil in range(n_coils):
        map(mapper, mapping, f"magnetics/coil[{coil}]/name")
        num_positions = map(mapper, mapping, f"magnetics/coil[{coil}]/position")
        for position in range(num_positions):
            map(mapper, mapping, f"magnetics/coil[{coil}]/position[{position}]/r")
            map(mapper, mapping, f"magnetics/coil[{coil}]/position[{position}]/z")
        map(mapper, mapping, f"magnetics/coil[{coil}]/flux/time")
        map(mapper, mapping, f"magnetics/coil[{coil}]/flux/data")
        map(mapper, mapping, f"magnetics/coil[{coil}]/flux/dot_product")


def main(args):
    cxxlibs = False
    if len(args) == 2:
        if args[1] == "--help":
            print(f"Usage: python {args[0]} [--cxxlibs]")
            sys.exit(0)
        elif args[1] == "--cxxlibs":
            cxxlibs = True
        else:
            print(f"Usage: python {args[0]} [--cxxlibs]")
            sys.exit(1)

    print("Calling LibTokaMap version:", libtokamap.__version__)

    root = Path().absolute().parent
    build_root = root / "build" / "examples" / "simple_mapper"

    mapping_directory = root / "examples" / "simple_mapper" / "mappings"
    mapper = libtokamap.Mapper(str(mapping_directory))

    data_root = root / "examples" / "simple_mapper" / "data"

    if cxxlibs:
        build_root = root / "build" / "examples" / "simple_mapper"
        factory_library = build_root / ("libjson_data_source" + libtokamap.LibrarySuffix)
        mapper.register_data_source_factory("JSONFactory", str(factory_library))
        mapper.register_data_source("JSON", "JSONFactory", {"data_root": data_root})
        custom_function_library = build_root / ("libcustom_library" + libtokamap.LibrarySuffix)
        mapper.load_custom_function_library(custom_function_library)
    else:
        mapper.register_python_data_source("JSON", JSONDataSource(data_root))
        mapper.register_custom_function("custom", "dot_product", dot_product)

    mapping = "EXAMPLE"
    try:
        map_all(mapper, mapping)
    except Exception as e:
        print(f"{e}")


if __name__ == "__main__":
    import sys
    import timeit
    print(timeit.timeit("main(sys.argv)", number=10, setup="from __main__ import main"))

File: examples/simple_mapper/src/simple_mapper.cpp
#include <algorithm>
#include <bit>
#include <cstddef>
#include <cstdint>
#include <exception>
#include <filesystem>
#include <iostream>
#include <libtokamap.hpp>
#include <nlohmann/json.hpp>
#include <string>
#include <typeindex>
#include <vector>

namespace
{

uint64_t map(libtokamap::MappingHandler& mapping_handler, const std::string& mapping, const std::string& path)
{
    std::type_index data_type = std::type_index{typeid(char)};
    int rank = 1;
    nlohmann::json extra_attributes = {{"shot", 42}};

    auto result = mapping_handler.map(mapping, path, data_type, rank, extra_attributes);

    constexpr int64_t key_length = 35;
    std::string padding;
    padding.resize(std::max(key_length - static_cast<int64_t>(path.size()), int64_t{0}), ' ');
    std::cout << path << padding << " -> " << result.to_string() << "\n";

    if (result.type_index() == std::type_index{typeid(uint64_t)}) {
        return *std::bit_cast<uint64_t*>(result.buffer());
    }
    return 0;
}

void map_all(libtokamap::MappingHandler& mapping_handler, const std::string& mapping)
{
    map(mapping_handler, mapping, "magnetics/version");
    auto num_coils = map(mapping_handler, mapping, "magnetics/coil");
    for (auto coil = 0UL; coil < num_coils; ++coil) {
        std::string path = "magnetics/coil[" + std::to_string(coil) + "]";
        map(mapping_handler, mapping, path + "/name");
        auto num_positions = map(mapping_handler, mapping, path + "/position");
        for (auto position = 0UL; position < num_positions; ++position) {
            std::string pos_path = path + ("/position[" + std::to_string(position) + "]");
            map(mapping_handler, mapping, pos_path + "/r");
            map(mapping_handler, mapping, pos_path + "/z");
        }
        map(mapping_handler, mapping, path + "/flux/time");
        map(mapping_handler, mapping, path + "/flux/data");
        map(mapping_handler, mapping, path + "/flux/dot_product");
    }
}

} // namespace

int main()
{
    try {
        libtokamap::MappingHandler mapping_handler;

        auto root = std::filesystem::path{__FILE__}.parent_path().parent_path();
        auto build_root = root.parent_path().parent_path() / "build" / "examples" / "simple_mapper";

        auto data_source_library = build_root / (std::string{"libjson_data_source"} + libtokamap::LibrarySuffix);
        mapping_handler.register_data_source_factory("JSON_factory", data_source_library);

        std::filesystem::path data_root = root / "data";
        libtokamap::DataSourceFactoryArgs factory_args = {{"data_root", data_root}};
        mapping_handler.register_data_source("JSON", "JSON_factory", factory_args);

        std::string library_name = std::string{"libcustom_library"} + libtokamap::LibrarySuffix;
        std::vector<std::string> custom_function_libraries = {build_root / library_name};

        auto schema_root = root.parent_path().parent_path() / "schemas";
        nlohmann::json config = {{"mapping_directory", (root / "mappings").string()},
                                 {"schemas_directory", schema_root.string()},
                                 {"custom_function_libraries", custom_function_libraries}};
        mapping_handler.init(config);

        const char* mapping = "EXAMPLE";
        map_all(mapping_handler, mapping);
    } catch (std::exception& ex) {
        std::cerr << "error: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}
